File: WebClient SDK/src/default/ui/cardFiles/initializeFilesViewUIExtension.ts
import { FilesViewGeneratorBaseUIExtension } from './filesViewGeneratorBaseUIExtension';
import { FileControlCreationParams } from './fileControlCreationParams';
import { FilesViewCardControlInitializationStrategy } from './filesViewCardControlInitializationStrategy';
import { FilesViewControlContentItemsFactory } from './filesViewControlContentItemsFactory';
import { ICardModel, ICardUIExtensionContext } from 'tessa/ui/cards';
import { executeExtensions, CardTypeExtensionTypes, TypeExtensionContext } from 'tessa/cards';
import { tryGetFromSettings } from 'tessa/ui';
import { DefaultFormTabWithTasksViewModel } from 'tessa/ui/cards/forms';
import {
  FileListViewModel,
  GridRowAction,
  GridRowEventArgs,
  GridViewModel,
  IViewControlInitializationStrategy,
  ViewControlViewModel
} from 'tessa/ui/cards/controls';
import { TypeSettingsSealed } from 'tessa/cards/types';
import { IStorage } from 'tessa/platform/storage';
import { tryGetFileViewExtensionInitializationStrategyHandlers } from './cardFilesExtensions';

export class InitializeFilesViewUIExtension extends FilesViewGeneratorBaseUIExtension {
  private _disposes: Array<(() => void) | null> = [];

  async initializing(context: ICardUIExtensionContext): Promise<void> {
    const result = await executeExtensions(
      CardTypeExtensionTypes.initializeFilesView,
      context.card,
      context.model.generalMetadata,
      this.executeInitializingAction,
      context
    );

    context.validationResult.add(result);
  }

  async initialized(context: ICardUIExtensionContext): Promise<void> {
    const result = await executeExtensions(
      CardTypeExtensionTypes.initializeFilesView,
      context.card,
      context.model.generalMetadata,
      this.executeInitializedAction,
      context
    );

    context.validationResult.add(result);
  }

  finalized(): void {
    for (const dispose of this._disposes) {
      if (dispose) {
        dispose();
      }
    }
    this._disposes.length = 0;
  }

  private executeInitializingAction = async (context: TypeExtensionContext) => {
    const settings = context.settings;
    const filesViewAlias = tryGetFromSettings<string>(settings, 'FilesViewAlias', '');

    const options = new FileControlCreationParams();
    options.categoriesViewAlias = tryGetFromSettings<string>(settings, 'CategoriesViewAlias', '');
    options.previewControlName = tryGetFromSettings<string>(settings, 'PreviewControlName', '');
    options.isCategoriesEnabled = tryGetFromSettings<boolean>(
      settings,
      'IsCategoriesEnabled',
      false
    );
    options.isIgnoreExistingCategories = tryGetFromSettings<boolean>(
      settings,
      'IsIgnoreExistingCategories',
      false
    );
    options.isManualCategoriesCreationDisabled = tryGetFromSettings<boolean>(
      settings,
      'IsManualCategoriesCreationDisabled',
      false
    );
    options.isNullCategoryCreationDisabled = tryGetFromSettings<boolean>(
      settings,
      'IsNullCategoryCreationDisabled',
      false
    );
    options.categoriesViewMapping = tryGetFromSettings<IStorage[]>(
      settings,
      'CategoriesViewMapping',
      []
    );

    const uiContext = context.externalContext as ICardUIExtensionContext;
    const cardTask = context.cardTask;
    if (!cardTask) {
      await this.initializeFileControl(uiContext.model, filesViewAlias, options);
    } else {
      uiContext.model.taskInitializers.push(async taskCardModel => {
        if (taskCardModel.cardTask === cardTask) {
          await this.initializeFileControl(taskCardModel, filesViewAlias, options);
        }
      });
    }
  };

  private executeInitializedAction = async (context: TypeExtensionContext) => {
    const uiContext = context.externalContext as ICardUIExtensionContext;
    const settings = context.settings;
    if (!context.cardTask) {
      const fileControl = await this.attachViewToFileControlCore(context, uiContext.model);
      this.initializeDefaultGroupInFileControl(fileControl, settings);
      if (!fileControl) {
        const tables = uiContext.model.controlsBag.filter(
          x => x instanceof GridViewModel
        ) as GridViewModel[];
        for (const table of tables) {
          this._disposes.push(
            table.rowInvoked.addWithDispose(e => {
              if (e.action == GridRowAction.Inserted || e.action == GridRowAction.Opening) {
                this.initializeExtensionForTableForm(context, e);
              }
            })
          );
        }
      }
    } else {
      const model = uiContext.model;
      const tasks = (model.mainForm as DefaultFormTabWithTasksViewModel).tasks;
      if (!tasks) {
        return;
      }
      const task = tasks.find(x => x.taskModel.cardTask === context.cardTask);
      if (task) {
        task.modifyWorkspace(async () => {
          const fileControl = await this.attachViewToFileControlCore(context, task.taskModel);
          this.initializeDefaultGroupInFileControl(fileControl, settings);
          if (!fileControl) {
            const tables = task.taskModel.controlsBag.filter(
              x => x instanceof GridViewModel
            ) as GridViewModel[];
            for (const table of tables) {
              this._disposes.push(
                table.rowInvoked.addWithDispose(e => {
                  if (e.action == GridRowAction.Inserted || e.action == GridRowAction.Opening) {
                    this.initializeExtensionForTableForm(context, e);
                  }
                })
              );
            }
          }
        });
      }
    }
  };

  private initializeDefaultGroupInFileControl(
    fileControl: FileListViewModel | null,
    settings: TypeSettingsSealed | null
  ) {
    if (fileControl) {
      const defaultGroup = tryGetFromSettings<string>(settings, 'DefaultGroup', '');
      if (defaultGroup) {
        fileControl.selectedGrouping =
          fileControl.groupings.find(x => x.name === defaultGroup) ?? null;
      }
    }
  }

  private initializeExtensionForTableForm(context: TypeExtensionContext, e: GridRowEventArgs) {
    const cardViewControlViewModels = e.rowModel!.controlsBag.filter(
      x => x instanceof ViewControlViewModel
    ) as ViewControlViewModel[];
    if (cardViewControlViewModels.length > 0) {
      const fileControl = this.attachViewToFileControlCore(context, e.rowModel!);
      if (!fileControl) {
        const tables = e.rowModel!.controlsBag.filter(
          x => x instanceof GridViewModel
        ) as GridViewModel[];
        for (const table of tables) {
          this._disposes.push(
            table.rowInvoked.addWithDispose(e => {
              if (e.action == GridRowAction.Inserted || e.action == GridRowAction.Opening) {
                this.initializeExtensionForTableForm(context, e);
              }
            })
          );
        }
      }
    }
  }

  private async attachViewToFileControlCore(
    extensionContext: TypeExtensionContext,
    cardModel: ICardModel
  ): Promise<FileListViewModel | null> {
    let strategy: IViewControlInitializationStrategy | null = null;
    for (const handler of tryGetFileViewExtensionInitializationStrategyHandlers(cardModel) ?? []) {
      strategy = handler(extensionContext, cardModel);
      if (!!strategy) {
        break;
      }
    }
    const settings = extensionContext.settings;
    const filesViewAlias = tryGetFromSettings<string>(settings, 'FilesViewAlias', '');
    return await this.attachViewToFileControl(
      cardModel,
      filesViewAlias,
      strategy ??
        new FilesViewCardControlInitializationStrategy(new FilesViewControlContentItemsFactory())
    );
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/CardFiles/InitializeFilesViewUIExtension.cs
#nullable enable
using System.Collections;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Cards;
using Tessa.Extensions.Default.Shared.Views;
using Tessa.Platform;
using Tessa.Platform.Runtime;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;
using Tessa.UI.Files;
using Tessa.UI.Views.Content;
using Tessa.Views;
using Unity;

namespace Tessa.Extensions.Default.Client.UI.CardFiles
{
    /// <summary>
    /// Реализация расширения типа карточки для преобразования представления в файловый контрол
    /// </summary>
    public sealed class InitializeFilesViewUIExtension : FilesViewGeneratorBaseUIExtension
    {
        #region Private fields

        private readonly ICardMetadata cardMetadata;

        private readonly IViewCardControlInitializationStrategy initializationStrategy;

        #endregion

        #region Constructor

        public InitializeFilesViewUIExtension(
            ISession session,
            IExtensionContainer extensionContainer,
            IViewService viewService,
            ICardMetadata cardMetadata,
            IProcessNameResolver processNameResolver,
            [Dependency(nameof(FilesViewCardControlInitializationStrategy))] IViewCardControlInitializationStrategy initializationStrategy)
            : base(session, extensionContainer, viewService, processNameResolver)
        {
            this.cardMetadata = NotNullOrThrow(cardMetadata);
            this.initializationStrategy = NotNullOrThrow(initializationStrategy);
        }

        #endregion

        #region Private methods

        private async Task ExecuteInitializingActionAsync(ITypeExtensionContext context)
        {
            var settings = NotNullOrThrow(context.Settings);
            var filesViewAlias = settings.TryGet<string>(DefaultCardTypeExtensionSettings.FilesViewAlias);

            var options = new FileControlCreationParams
            {
                CategoriesViewAlias = settings.TryGet<string>(DefaultCardTypeExtensionSettings.CategoriesViewAlias),
                PreviewControlName = settings.TryGet<string>(DefaultCardTypeExtensionSettings.PreviewControlName),
                IsCategoriesEnabled = settings.TryGet<bool>(DefaultCardTypeExtensionSettings.IsCategoriesEnabled),
                IsIgnoreExistingCategories =
                    settings.TryGet<bool>(DefaultCardTypeExtensionSettings.IsIgnoreExistingCategories),
                IsManualCategoriesCreationDisabled =
                    settings.TryGet<bool>(DefaultCardTypeExtensionSettings.IsManualCategoriesCreationDisabled),
                IsNullCategoryCreationDisabled =
                    settings.TryGet<bool>(DefaultCardTypeExtensionSettings.IsNullCategoryCreationDisabled),
                CategoriesViewMapping = settings.TryGet<IList>(DefaultCardTypeExtensionSettings.CategoriesViewMapping)
            };
            var uiContext = (ICardUIExtensionContext) NotNullOrThrow(context.ExternalContext);

            if (context.CardTask is null)
            {
                this.AddCardModelInitializers(uiContext.Model, filesViewAlias, options);
            }
            else
            {
                uiContext.Model.TaskInitializers.Add(async (taskCardModel, ct) =>
                {
                    if (taskCardModel.CardTask == context.CardTask)
                    {
                        this.AddCardModelInitializers(taskCardModel, filesViewAlias, options);
                    }
                });
            }
        }

        private async Task ExecuteInitializedActionAsync(ITypeExtensionContext context)
        {
            var uiContext = (ICardUIExtensionContext) NotNullOrThrow(context.ExternalContext);

            if (context.CardTask is null)
            {
                var isAttached = await this.AttachViewToFileControlCoreAsync(
                    context,
                    uiContext.Model,
                    context.CancellationToken);
                if (!isAttached)
                {
                    var tables = uiContext.Model.ControlBag.OfType<GridViewModel>();
                    foreach (var table in tables)
                    {
                        table.RowInvoked += (s, e) =>
                        {
                            if (e.Action is GridRowAction.Inserted or GridRowAction.Opening)
                            {
                                this.InitializeExtensionForTableForm(context, e);
                            }
                        };
                    }
                }
            }
            else
            {
                await uiContext.Model.ModifyTasksAsync(async (task, model) =>
                {
                    if (task.TaskModel.CardTask == context.CardTask)
                    {
                        await task.ModifyWorkspaceAsync(async (t, subscribeToTaskModel) =>
                        {
                            var isAttached = await this.AttachViewToFileControlCoreAsync(
                                context,
                                task.TaskModel,
                                CancellationToken.None);
                            if (!isAttached)
                            {
                                var tables = task.TaskModel.ControlBag.OfType<GridViewModel>();
                                foreach (var table in tables)
                                {
                                    table.RowInvoked += (s, e) =>
                                    {
                                        if (e.Action is GridRowAction.Inserted or GridRowAction.Opening)
                                        {
                                            this.InitializeExtensionForTableForm(context, e);
                                        }
                                    };
                                }
                            }
                        });
                    }
                });
            }
        }

        private void InitializeExtensionForTableForm(ITypeExtensionContext context, RowEventArgs e)
        {
            var cardViewControlViewModels = e.RowModel.ControlBag.OfType<CardViewControlViewModel>();
            if (cardViewControlViewModels.Any())
            {
                DispatcherHelper.InvokeInUI(async () =>
                {
                    var isAttached = await this.AttachViewToFileControlCoreAsync(
                        context,
                        e.RowModel,
                        CancellationToken.None);
                    if (!isAttached)
                    {
                        var tables = e.RowModel.ControlBag.OfType<GridViewModel>();
                        foreach (var table in tables)
                        {
                            table.RowInvoked += (s2, e2) => { this.InitializeExtensionForTableForm(context, e2); };
                        }
                    }
                });
            }
        }

        private async ValueTask<bool> AttachViewToFileControlCoreAsync(
            ITypeExtensionContext extensionContext,
            ICardModel cardModel,
            CancellationToken cancellationToken = default)
        {
            IViewCardControlInitializationStrategy? strategy = null;
            foreach (var handlerAsync in cardModel.TryGetFileViewExtensionInitializationStrategyHandlers()
                     ?? Enumerable.Empty<TryGetControlInitializationStrategyAsync>())
            {
                strategy = await handlerAsync(extensionContext, cardModel, cancellationToken);
                if (strategy is not null)
                {
                    break;
                }
            }

            return await AttachViewToFileControlAsync(
                cardModel,
                extensionContext.Settings,
                CreateRowFunc,
                strategy ?? this.initializationStrategy,
                cancellationToken: cancellationToken);
        }
        
        private static ViewControlRowViewModel CreateRowFunc(TableRowCreationOptions options)
        {
            var fileViewModel = (IFileViewModel) options.Data[ColumnsConst.FileViewModel];
            options.Data.Remove(ColumnsConst.FileViewModel);

            return new TableFileRowViewModel(fileViewModel, options)
            {
                AutomationId = $"{fileViewModel.Model.ID}",
                AutomationName = fileViewModel.Model.Name,
            };
        }

        #endregion

        #region Base overrides

        public override async Task Initializing(ICardUIExtensionContext context)
        {
            ValidationResult result = await CardHelper
                .ExecuteTypeExtensionsAsync(
                    DefaultCardTypeExtensionTypes.InitializeFilesView,
                    context.Card,
                    this.cardMetadata,
                    this.ExecuteInitializingActionAsync,
                    context,
                    cancellationToken: context.CancellationToken);

            context.ValidationResult.Add(result);
        }

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            // Вот здесь проходит основная инициализация. Она не запускается при открытии мелкой карточки.
            ValidationResult result = await CardHelper
                .ExecuteTypeExtensionsAsync(
                    DefaultCardTypeExtensionTypes.InitializeFilesView,
                    context.Card,
                    this.cardMetadata,
                    this.ExecuteInitializedActionAsync,
                    context);

            context.ValidationResult.Add(result);
        }

        #endregion
    }
}
